### PR TITLE
Add and show empty folders

### PR DESCRIPTION
**Pull request by MancunianSam**
This retrieves empty folders and constructs the path string for them and returns them in an empty folders object.
It also takes an empty files argument and uses this to add entries to the File table.

### CODE REVIEW

File: src/main/scala/uk/gov/nationalarchives/tdr/api/http/GraphQLServer.scala [hunk 1]
```diff
@@ -31,6 +31,7 @@ class GraphQLServer(slickSession: SlickSession) {
   private val logger = Logger(s"${GraphQLServer.getClass}")
   private val config = ConfigFactory.load()
 
+
```
Comment (TomJKing) [inline src/main/scala/uk/gov/nationalarchives/tdr/api/http/GraphQLServer.scala:34]: Unnecessary line break

File: src/test/resources/json/getconsignment_data_files_and_folders.json [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+          "fileId": "6420152a-aaf2-4401-a309-f67ae35f5702",
+          "fileType": "File",
+          "metadata": {
+            "clientSideOriginalFilePath": "ClientSideOriginalFilepath value"
+          }
+        }
+      ]
+    }
+  }
+}
```
Comment (TomJKing) [inline src/test/resources/json/getconsignment_data_files_and_folders.json:29]: Missing line break

File: src/test/resources/json/getconsignment_query_file_data.json [hunk 3]
```diff
@@ -0,0 +1,3 @@
+{
+  "query": "{getConsignment(consignmentid: \"e72d94d5-ae79-4a05-bee9-86d9dea2bcc9\"){files {fileId fileType metadata {clientSideOriginalFilePath}}}}"
+}
```
Comment (TomJKing) [inline src/test/resources/json/getconsignment_query_file_data.json:3]: Missing line break

File: src/test/scala/uk/gov/nationalarchives/tdr/api/routes/ConsignmentRouteSpec.scala [hunk 4]
```diff
@@ -390,6 +390,24 @@ class ConsignmentRouteSpec extends TestContainerUtils with Matchers with TestReq
...
+      utils.createConsignment(consignmentId, userId)
+      staticMetadataProperties.map(_.name).foreach(utils.addFileProperty)
+      clientSideProperties.foreach(utils.addFileProperty)
+      val topDirectory = UUID.fromString("ce0a51a5-a224-474f-b3a4-df75effd5b34")
+      val subDirectory = UUID.fromString("2753ceca-4df3-436b-8891-78ad38e2e8c5")
+      val fileId = UUID.fromString("6420152a-aaf2-4401-a309-f67ae35f5702")
+      createDirectoryAndMetadata(consignmentId, topDirectory, utils, "directory")
+      createDirectoryAndMetadata(consignmentId, subDirectory, utils, "subDirectory", topDirectory.some)
+      setUpFileAndStandardMetadata(consignmentId, fileId, utils, subDirectory.some)
+      val expectedResponse: GraphqlQueryData = expectedQueryResponse("data_files_and_folders")
```
Comment (TomJKing) [inline src/test/scala/uk/gov/nationalarchives/tdr/api/routes/ConsignmentRouteSpec.scala:406]: Maybe a line break here, just to split out the test setup from from the actual testing?

File: src/main/scala/uk/gov/nationalarchives/tdr/api/graphql/fields/FileFields.scala [hunk 5]
```diff
@@ -1,14 +1,15 @@
...
 import io.circe.generic.auto._
 import sangria.macros.derive._
 import sangria.marshalling.circe._
 import sangria.schema.{Argument, Field, InputObjectType, ListType, ObjectType, fields}
-import uk.gov.nationalarchives.tdr.api.auth.{ValidateHasExportAccess, ValidateUserHasAccessToConsignment}
+import uk.gov.nationalarchives.tdr.api.auth.ValidateUserHasAccessToConsignment
 import uk.gov.nationalarchives.tdr.api.graphql.ConsignmentApiContext
 import uk.gov.nationalarchives.tdr.api.graphql.validation.UserOwnsConsignment
-import uk.gov.nationalarchives.tdr.api.graphql.fields.FieldTypes._
+import uk.gov.nationalarchives.tdr.api.graphql.fields.FieldTypes.UuidType
```
Comment (TomJKing) [inline src/main/scala/uk/gov/nationalarchives/tdr/api/graphql/fields/FileFields.scala:10]: Think this is an unnecessary import: `import uk.gov.nationalarchives.tdr.api.graphql.fields.FieldTypes.UuidType`

File: src/main/scala/uk/gov/nationalarchives/tdr/api/model/file/NodeType.scala [hunk 6]
```diff
@@ -15,14 +15,14 @@ object NodeType {
       if (deconstructed.length == 1) None else Some(deconstructed.dropRight(1).mkString("/"))
 
     def fileType: String = {
```
Comment (TomJKing) [inline src/main/scala/uk/gov/nationalarchives/tdr/api/model/file/NodeType.scala:17]: Maybe rename this to: `typeIdentifier` to make it more consistent, as not strictly a "file type" anymore?

File: src/main/scala/uk/gov/nationalarchives/tdr/api/service/FileService.scala [hunk 7]
```diff
@@ -1,18 +1,17 @@
 package uk.gov.nationalarchives.tdr.api.service
 
-import java.sql.Timestamp
-import java.util.UUID
-
 import uk.gov.nationalarchives.Tables.{FileRow, FilemetadataRow}
 import uk.gov.nationalarchives.tdr.api.db.repository._
 import uk.gov.nationalarchives.tdr.api.graphql.fields.FileFields.{AddFileAndMetadataInput, ClientSideMetadataInput, FileMatches}
 import uk.gov.nationalarchives.tdr.api.model.file.NodeType
```
Comment (TomJKing) [inline src/main/scala/uk/gov/nationalarchives/tdr/api/service/FileService.scala:6]: Don't think this import is needed anymore: `import uk.gov.nationalarchives.tdr.api.model.file.NodeType`

File: src/main/scala/uk/gov/nationalarchives/tdr/api/service/FileService.scala [hunk 8]
```diff
@@ -40,11 +39,15 @@ class FileService(
     val now = Timestamp.from(timeSource.now)
     val consignmentId = addFileAndMetadataInput.consignmentId
     val filePaths = addFileAndMetadataInput.metadataInput.map(_.originalPath).toSet
-    val allNodes: Map[String, TreeNode] = treeNodesUtils.generateNodes(filePaths)
+    val allFileNodes: Map[String, TreeNode] = treeNodesUtils.generateNodes(filePaths, fileTypeIdentifier)
+    val allEmptyDirectoryNodes: Map[String, TreeNode] = treeNodesUtils.generateNodes(addFileAndMetadataInput.emptyDirectories.toSet, directoryTypeIdentifier)
+
+    val folderIdTopath: Map[UUID, String] = (allEmptyDirectoryNodes ++ allFileNodes.filter(_._2.treeNodeType == directoryTypeIdentifier))
```
Comment (TomJKing) [inline src/main/scala/uk/gov/nationalarchives/tdr/api/service/FileService.scala:45]: Rename to camel case: `folderIdToPath`

File: src/main/scala/uk/gov/nationalarchives/tdr/api/service/FileService.scala [hunk 9]
```diff
@@ -63,7 +66,10 @@ class FileService(
           row(fileId, property.value, property.name)
         })
       case _ => Seq()
-    }
+    } ++ fileRows.filter(_.filetype.get.isDirectoryType).flatMap(folder => {
```
Comment (TomJKing) [inline src/main/scala/uk/gov/nationalarchives/tdr/api/service/FileService.scala:69]: Maybe rename `folder` to `directory` or `dir` for consistency?

File: src/main/scala/uk/gov/nationalarchives/tdr/api/utils/TreeNodesUtils.scala [hunk 10]
```diff
@@ -1,32 +1,32 @@
...
-      val name = jioFile.getName
-      val treeNode = TreeNode(uuidSource.uuid, name, parentPath, fileType)
-      val nextMap = nodes + (originalPath -> treeNode)
-      if (parentPath.isEmpty) {
-        nextMap
-      } else {
-        innerFunction(parentPath.get, folderTypeIdentifier, nextMap)
-      }
+  @tailrec
+  private def innerFunction(originalPath: String, fileType: String, nodes: Map[String, TreeNode]): Map[String, TreeNode] = {
```
Comment (TomJKing) [inline src/main/scala/uk/gov/nationalarchives/tdr/api/utils/TreeNodesUtils.scala:13]: Maybe change the `fileType` parameter to something else? Not sure what though?

File: src/test/scala/uk/gov/nationalarchives/tdr/api/routes/ConsignmentRouteSpec.scala [hunk 11]
```diff
@@ -390,6 +390,24 @@ class ConsignmentRouteSpec extends TestContainerUtils with Matchers with TestReq
       response.errors.head.message should equal(expectedResponse.errors.head.message)
   }
 
+  "getConsignment" should "return directories in the files list" in withContainers {
```
Comment (TomJKing) [inline src/test/scala/uk/gov/nationalarchives/tdr/api/routes/ConsignmentRouteSpec.scala:393]: Maybe rename test slightly to: `return files and directories in the files list`

File: src/test/resources/json/getconsignment_data_files_and_folders.json [hunk 12]
```diff
@@ -0,0 +1,29 @@
+{
```
Comment (TomJKing) [inline src/test/resources/json/getconsignment_data_files_and_folders.json:1]: Maybe rename file to: `.../getconsignment_data_files_and_directories.json` ?

File: src/test/scala/uk/gov/nationalarchives/tdr/api/routes/FileRouteSpec.scala [hunk 13]
```diff
@@ -50,6 +50,24 @@ class FileRouteSpec extends TestContainerUtils with Matchers with TestRequest {
...
   }
 
+  "The api" should "add files and metadata entries for files and directories" in withContainers {
+    case container: PostgreSQLContainer =>
+      val consignmentId = UUID.fromString("f1a9269d-157b-402c-98d8-1633393634c5")
+      val utils = TestUtils(container.database)
+      (clientSideProperties ++ staticMetadataProperties.map(_.name)).foreach(utils.addFileProperty)
+      utils.createConsignment(consignmentId, userId)
+
+      val response = runTestMutationFileMetadata("mutation_alldata", validUserToken())
```
Comment (TomJKing) [inline src/test/scala/uk/gov/nationalarchives/tdr/api/routes/FileRouteSpec.scala:60]: The val `response` is no longer used, as switched to checking with counts? Was there a reason for no longer using as JSON response to check the expected result?

File: src/test/scala/uk/gov/nationalarchives/tdr/api/routes/FileRouteSpec.scala [hunk 14]
```diff
@@ -50,6 +50,24 @@ class FileRouteSpec extends TestContainerUtils with Matchers with TestRequest {
...
+      val distinctDirectoryCount = 3
+      val fileCount = 2
+      val expectedCount = (staticMetadataProperties.size * distinctDirectoryCount) + //Directories have static metadata
```
Comment (TomJKing) [inline src/test/scala/uk/gov/nationalarchives/tdr/api/routes/FileRouteSpec.scala:63]: Are these comments necessary?

File: src/test/scala/uk/gov/nationalarchives/tdr/api/utils/FixedUUIDSource.scala [hunk 15]
```diff
@@ -5,35 +5,35 @@ import java.util.UUID
...
+        UUID.fromString("8b5bb406-5f12-11eb-ae93-0242ac130002"),
+        UUID.fromString("8b5bb546-5f12-11eb-ae93-0242ac130002"),
+        UUID.fromString("8b5bb708-5f12-11eb-ae93-0242ac130002"),
+        UUID.fromString("8b5bb898-5f12-11eb-ae93-0242ac130002"),
+        UUID.fromString("8b5bba6e-5f12-11eb-ae93-0242ac130002"),
+        UUID.fromString("8b5bbc44-5f12-11eb-ae93-0242ac130002"),
+        UUID.fromString("8b5bbd98-5f12-11eb-ae93-0242ac130002")
+      )(idx)
+    }
+
```
Comment (TomJKing) [inline src/test/scala/uk/gov/nationalarchives/tdr/api/utils/FixedUUIDSource.scala:36]: Unnecessary line break.

**Review by TomJKing — COMMENTED** [review]
Looks good. Just some minor comments and questions.

File: src/test/resources/json/getconsignment_query_file_data.json (hunk 3 above)
Comment (MancunianSam) [reply src/test/resources/json/getconsignment_query_file_data.json:3]: Sorry, I thought I'd checked for these.

File: src/main/scala/uk/gov/nationalarchives/tdr/api/graphql/fields/FileFields.scala (hunk 5 above)
Comment (MancunianSam) [reply src/main/scala/uk/gov/nationalarchives/tdr/api/graphql/fields/FileFields.scala:10]: I've removed it.

File: src/main/scala/uk/gov/nationalarchives/tdr/api/model/file/NodeType.scala (hunk 6 above)
Comment (MancunianSam) [reply src/main/scala/uk/gov/nationalarchives/tdr/api/model/file/NodeType.scala:17]: Makes sense, I've changed it.

File: src/main/scala/uk/gov/nationalarchives/tdr/api/service/FileService.scala (hunk 7 above)
Comment (MancunianSam) [reply src/main/scala/uk/gov/nationalarchives/tdr/api/service/FileService.scala:6]: I've deleted it, cheers.

File: src/main/scala/uk/gov/nationalarchives/tdr/api/service/FileService.scala (hunk 8 above)
Comment (MancunianSam) [reply src/main/scala/uk/gov/nationalarchives/tdr/api/service/FileService.scala:45]: I knew I'd messed up the case somewhere, I couldn't find it though.

File: src/main/scala/uk/gov/nationalarchives/tdr/api/utils/TreeNodesUtils.scala (hunk 10 above)
Comment (MancunianSam) [reply src/main/scala/uk/gov/nationalarchives/tdr/api/utils/TreeNodesUtils.scala:13]: I've called it `typeIdentifier` to match what the calling method calls it.

File: src/test/scala/uk/gov/nationalarchives/tdr/api/routes/FileRouteSpec.scala (hunk 13 above)
Comment (MancunianSam) [reply src/test/scala/uk/gov/nationalarchives/tdr/api/routes/FileRouteSpec.scala:60]: I've taken it out. The test above it checks the response but the response is basic, just `{ "fileId": "...", "matchId": 1 }` so I added in this test to check that all the database updates are being done properly.

File: src/test/scala/uk/gov/nationalarchives/tdr/api/routes/FileRouteSpec.scala (hunk 14 above)
Comment (MancunianSam) [reply src/test/scala/uk/gov/nationalarchives/tdr/api/routes/FileRouteSpec.scala:63]: They helped me to count :smile: They're not needed now, I've taken them out.

File: src/main/scala/uk/gov/nationalarchives/tdr/api/graphql/fields/FileFields.scala (hunk 5 above)
Comment (MancunianSam) [reply src/main/scala/uk/gov/nationalarchives/tdr/api/graphql/fields/FileFields.scala]: It is necessary, it's an implicit import that Intellij thinks isn't used but it won't compile without it. I've put it back.

Comment (MancunianSam) [reply src/main/scala/uk/gov/nationalarchives/tdr/api/graphql/fields/FileFields.scala]: It is necessary, it's an implicit import that Intellij thinks isn't used but it won't compile without it. I've put it back.

**Review by TomJKing — APPROVED** [review]
Looks good. Thanks

**Pull request merged**